backend/dal/Services/Admin/Concrete/AgencyService.cs

```diff
@@ -100,35 +100,133 @@ public Agency Get(int id)
+
+            // If the agency has become a sub-agency, or a parent-agency then users will need to be updated.
+            if (original.ParentId.HasValue && !agency.ParentId.HasValue)
+            {
+                // This agency has become a parent agency, all users associated with it will be removed
+                // because users should only be associated with a single parent.
+                updatedUsers.AddRange(this.Context.Users.Where(u => u.Agencies.Any(a => a.AgencyId == original.ParentId)));
+
+                // Remove any user who are linked to the agency.
+                var users = this.Context.Users.Include(u => u.Agencies).Where(u => u.Agencies.Any(ua => ua.AgencyId == agency.Id));
+                users.ForEach(u => u.Agencies.Clear());
+                this.Context.UpdateRange(users);
+            }
+            else if (!original.ParentId.HasValue && agency.ParentId.HasValue)
+            {
+                // This agency has become a sub-agency, all original users need their list of agencies reduced to only this agency.
+                updatedUsers.AddRange(this.Context.Users.Include(u => u.Agencies).Where(u => u.Agencies.Any(a => a.AgencyId == agency.ParentId)));
+            }
+            else if (original.ParentId != agency.ParentId)
+            {
+                // Remove the sub-agency from curretly linked users and add it to the users who belong to the new parent agency.
+                updatedUsers.AddRange(this.Context.Users.Where(u => u.Agencies.Any(a => a.AgencyId == original.ParentId)));
+                updatedUsers.AddRange(this.Context.Users.Where(u => u.Agencies.Any(a => a.AgencyId == agency.ParentId)));
+            }
 
-            var agency = this.Context.Agencies.Find(entity.Id) ?? throw new KeyNotFoundException();
+            if (original.IsDisabled != agency.IsDisabled)
+            {
+                if (agency.ParentId.HasValue && agency.IsDisabled)
+                {
+                    // Remove the sub-agency from users.
+                    updatedUsers.AddRange(this.Context.Users.Where(u => u.Agencies.Any(a => a.AgencyId == agency.ParentId)));
+                }
+                else if (!agency.ParentId.HasValue && agency.IsDisabled)
+                {
+                    // Remove the agency from the users.
+                    // This will result in the user not belonging to an agency.
+                    var users = this.Context.Users.Include(u => u.Agencies).Where(u => u.Agencies.Any(a => a.AgencyId == agency.Id));
+                    users.ForEach(u => u.Agencies.Clear());
+                    updatedUsers.AddRange(users);
+                    this.Context.UpdateRange(users);
+                }
+                else if (agency.ParentId.HasValue)
+                {
+                    // Add the sub-agency to users who are associated with the parent.
+                    updatedUsers.AddRange(this.Context.Users.Where(u => u.Agencies.Any(a => a.AgencyId == agency.ParentId)));
```

Comment:
yeah, so this code that is supposed to be doing the adding of the agency is identical to the code that does the removal in my above comment?

Reply:
Save answer as above.

Reply:
maybe just put them both in the same if statement then?